### ISSUE TITLE
Named exports

### ISSUE THREAD

**Issue by TrySound** [webpack-contrib/css-loader#509]
**Do you want to request a *feature* or report a *bug*?**

Feature 

**What is the current behavior?**

**What is the expected behavior?**

Would be great to have a named exports mode which can help to verify class existing by new webpack feature. Not quite sure it works now.  But in theory it can fix lots of is issues.

**Comment by TrySound** [comment]
Sorry
 Ux issues. Sitting in bar and t9

**Comment by michael-ciniawsky** [comment]
@TrySound 🍻 😛 You mean `import * as styles from 'a.css' && import {className, ...} from 'a.css` ?

**Comment by TrySound** [comment]
Yep. Didn't research this yet. Hope there will be valuable result.

**Comment by michael-ciniawsky** [comment]
Current plan is to remove `modules` completely from `css-loader` (webpack will support CSS out-of-the-box in the intermediate future) and move them to `postcss-loader` instead, this requires changes to the various `postcss-modules-*` plugins 

```js
postcss([ modules ]).process(css, options) // modules = [ scope, values, extract, ...] plugins
  .then((result) => {
     result.messages[i].modules = { 'className': '362zfef', ...} // support this (result.messages)
  })
```

```js
export default funtion loader () {
  ...
  postcss([ modules]).process(css, options)
    .then((result) => {
      if (let modules in result.messages) {
        return cb(null, result.css, result.map, { locals: modules })
      }

      return cb(null, result.css, result.map)
    })
    .catch((err) => cb(err))
}
```

+  \+ Exporting them as a ES2015 Module `export const locals = {...locals }` in `css-loader`

I && @evilebottnawi (Stylelint) will start working on that soon, chime anytime if interested 😛

**Comment by TrySound** [comment]
Should we wait for postcss 6 then? I think this stuff is not quite related.

**Comment by michael-ciniawsky** [comment]
Yep `postcss-loader` v2 is blocked bc of that atm, it's unrelated to the feature itself (named exports), but you mentioned, that you didn't researched anything yet, so that's the (in brief) Roadmap to named exports 😛, since in any case adding this is a breaking change CJS Export => ES2015 Export

**Comment by TrySound** [comment]
Here's example of config. `strictExportPresence` validates not existing exports of (!) harmony modules. So this module should contain at least one `import` or `export` statement. So yes, it's only webpack2 feature.

This config can fail with this error
```
module.exports = {
    entry: './src/index.js',
    output: {
        filename: 'dist/bundle.js'
    },
    module: {
        strictExportPresence: true
    }
};
```

```
ERROR in ./src/index.js
3:12-23 "export 'namedExport' was not found in './mod.js'
```

Empty output file may contains like this
```js
export const __cssModules__ = true;`
``

**Comment by evilebottnawi** [comment]
@TrySound can you create minimal reproduced test repo? and write in readme what do you have and what do expected
And yep `css-modules` will be dropped for `css-loader` in favor `plugin`s pack for `postcss-loader` or `css-modules-loader` (maybe best name), but if you we can resolve this issue now (maybe to other branch), we can migrate this code to `plugins-pack/css-modules-loader` and you get working enchantment asap after release

**Comment by TrySound** [comment]
# namedExports mode
It's all simple
```js
// index.js
import * as style from './myStyle.css';
console.log(style.myClass);
```
```
// before
.myClass {}

// after
export const __cssModules = true;
// ... some loading code
export const myClass = 'hash-123456';
```
or if css is empty just
```
export const __cssModules = true; // export to identify file as harmony contained and enable errors
```

**Comment by TrySound** [comment]
Did you started CSS modules loader project somewhere?

**Comment by evilebottnawi** [comment]
@TrySound https://github.com/webpack-contrib/css-loader/pull/442#issuecomment-297118489

**Comment by michael-ciniawsky** [comment]
@TrySound Not yet I plan to do so today evening (until tomorrow) aand come back here to notify when ready, @evilebottnawi (Stylelint, `css-loader`, `postcss-loader`) is also 'on board'  and if you want to jump in aswell, we plan to draft out a PostCSS Modules Plugin && `cssm-loader` prototype and then integrate the parts into where the belong in the existing `postcss-modules-*` plugins && `css-loader || postcss-loader`  😛

**Comment by michael-ciniawsky** [comment]
@TrySound @evilebottnawi  #402 has some goodies hidden aswell, the original author made PR's to (`style-loader`, `css-loader`, `extract-text-webpack-plugin` and  wrote a 'tree shaking' prototype (webpack Plugin)) also worth a look maybe

**Comment by michael-ciniawsky** [comment]
Better start from current `css-loader` && `postcss-modules` code base or from scratch? :stuck_out_tongue:

**Comment by TrySound** [comment]
I'd like to keep this project maintainable and do some breaking changes.

**Comment by evilebottnawi** [comment]
@TrySound after discussion we resolve `css-loader` should working only with `css`, not `css-modules` and etc stuff, `postcss-modules-*` written as `postcss` plugins and should be moved to external `loader` (with own pack and settings). This will facilitate the maintenance of the project and reduce a lot of issue not related to `css-loader` (your can see our issues, most of their related to `postcss-modules-*`).

**Comment by evilebottnawi** [comment]
Or related to working `css-loader` with `postcss-modules-*`, I think there are many people who need `css-modules` and we easy to find maintainers for this

**Comment by evilebottnawi** [comment]
@TrySound btw, your can be one of first champion of this :smile:

**Comment by danny-andrews** [comment]
Any news on when module support will be dropped from css-loader and where the alternative package will live? I'm very interested in helping out with getting named exports finished but it sounds like it should be done in another repo.

**Comment by evilebottnawi** [comment]
@danny-andrews in todo list

**Comment by danny-andrews** [comment]
@evilebottnawi I'm afraid I don't have the context to start the work on this. Was hoping there was somewhere existing to contribute. :)
Seems like it might also depend on this: https://github.com/webpack/webpack/pull/6448?

**Comment by evilebottnawi** [comment]
@danny-andrews abandoned, also `css-modules` should not be a part of css pipeline in webpack, a lot of people use vanilla css, or scss/less or css-in-js and etc. In `css-loader@1` we drop css-modules support inside loader, better use for this babel-plugin or `css-modules-loader` (need implement)
